File: 2ª Fase 2018/Competição/H_layoutlinear/solutions/layoutlinear.c
/*
╔══════════════════════════════════════════════════════════════════════════════╗
║ Instituição   :  Faculdade de Tecnologia de São Paulo                        ║
║ Departamento  :  Tecnologia da Informação                                    ║
║ Curso         :  Análise e Desenvolvimento de Sistemas                       ║
║ Autor         :  Lucio Nunes de Lira                                         ║
╠══════════════════════════════════════════════════════════════════════════════╣
║ Competição    :  InterFatecs 2018 - 2ª fase                                  ║
║ Programa      :  Problema H - Layout Linear                                  ║
║ Linguagem     :  C99                                                         ║
║ Compilador    :  Pelles C (8.00.60)                                          ║
║ Versão        :  A (Rev. 0)                                                  ║
╚══════════════════════════════════════════════════════════════════════════════╝
*/

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct componente {
    int altura;
    char esp_largura;
    int largura;
    char alinhamento;
    char conteudo[101];
} Componente;

int altura(Componente com[], int n) {
	int t = 0;
	for (int i = 0; i < n; i++) t += com[i].altura;
	return t;
}

int maximo(int a, int b) {
	return a > b ? a : b;
}

int tam_layout(Componente com[], int n, int t) {
	int m = 0;
	for (int i = 0; i < n; i++)
		if (com[i].esp_largura == 'A') m = maximo(m, com[i].largura);
 		else m = maximo(m, strlen(com[i].conteudo) * t);
	return m;
}

int main(void) {
    Componente com[100];
    char f, lixo[12], z, c[101];
    int n, t, a, l;
   	scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        scanf("%d%*c%c%*c", &a, &f);
        if (f == 'A')
            scanf("%d%*c", &l);
        else {
            scanf("%s%*c", lixo);
            l = lixo[0];
        }
        scanf("%c%*c", &z);
        gets(c);
        com[i].altura = a;
		com[i].esp_largura = f;
		com[i].largura = l;
		com[i].alinhamento = z;
		strcpy(com[i].conteudo, c);
    }
    scanf("%d", &t);

	printf("%d\n", altura(com, n));
	int tl = tam_layout(com, n, t);
	for (int i = 0; i < n; i++) {
		if (com[i].esp_largura == 'A') {
			int esp = tl - com[i].largura;
			switch (com[i].alinhamento) {
				case ('E'):
					printf("%d %d\n", 0, esp); break;
				case ('C'):
					printf("%d %d\n", esp / 2, esp / 2); break;
				default:
					printf("%d %d\n", esp, 0); break;
			}
		}
		else
			if (com[i].largura == 'W') {
				int esp = tl - strlen(com[i].conteudo) * t;
				switch (com[i].alinhamento) {
					case ('E'):
						printf("%d %d\n", 0, esp); break;
					case ('C'):
						printf("%d %d\n", esp / 2, esp / 2); break;
					default:
						printf("%d %d\n", esp, 0); break;
				}
			}
			else
				printf("0 0\n");
	}

    return 0;
}

File: 2ª Fase 2018/Competição/H_layoutlinear/solutions/layoutlinear.py
'''
╔══════════════════════════════════════════════════════════════════════════════╗
║ Instituição   :  Faculdade de Tecnologia de São Paulo                        ║
║ Departamento  :  Tecnologia da Informação                                    ║
║ Curso         :  Análise e Desenvolvimento de Sistemas                       ║
║ Autor         :  Lucio Nunes de Lira                                         ║
╠══════════════════════════════════════════════════════════════════════════════╣
║ Competição    :  InterFatecs 2018 - 2ª fase                                  ║
║ Programa      :  Problema H - Layout Linear                                  ║
║ Linguagem     :  Python 3                                                    ║
║ Compilador    :  CPython (3.6.4)                                             ║
║ Versão        :  A (Rev. 0)                                                  ║
╚══════════════════════════════════════════════════════════════════════════════╝
'''

class Componente:
    def __init__(self, alt, esp, lar, ali, con):
        self.altura = alt
        self.esp_largura = esp
        self.largura = lar
        self.alinhamento = ali
        self.conteudo = con

def altura(com):
    t = 0
    for c in com: t += c.altura
    return t

def tam_layout(com, t):
    m = 0
    for c in com:
        if c.esp_largura == 'A': m = max(m, c.largura)
        else: m = max(m, len(c.conteudo)*t)
    return m

def separa(linha, i):
    while linha[i] != ' ': i += 1
    return i

n = int(input())
com = []
for i in range(n):
    linha = input()
    ini = 0
    fim = separa(linha, ini)
    a = int(linha[ini:fim])
    ini = fim+1
    fim = separa(linha, ini)
    f = linha[ini:fim]
    ini = fim + 1
    fim = separa(linha, ini)
    if f == 'A':
        l = int(linha[ini:fim])
    else:
        l = linha[ini:fim]
    ini = fim + 1
    fim = separa(linha, ini)
    z = linha[ini:fim]
    c = linha[fim+1:]
    com.append(Componente(a, f, l, z, c))
t = int(input())

print(altura(com))
tl = tam_layout(com, t)
for c in com:
    if c.esp_largura == 'A':
        esp = tl - c.largura
        if c.alinhamento == 'E':
            print('%d %d' % (0, esp))
        elif c.alinhamento == 'C':
            print('%d %d' % (esp // 2, esp // 2))
        else:
            print('%d %d' % (esp, 0))
    else:
        if c.largura == 'WRAPCONTENT':
            esp = tl - len(c.conteudo) * t
            if c.alinhamento == 'E':
                print('%d %d' % (0, esp))
            elif c.alinhamento == 'C':
                print('%d %d' % (esp // 2, esp // 2))
            else:
                print('%d %d' % (esp, 0))
        else:
            print('0 0')
